File: HomeAutomation/HomeAutomation/src/Components/TemperatureController.java
package Components;

import Core.Observer;
import Core.SmartHomeData;

public class TemperatureController implements Observer {
    private final SmartHomeData homeData;
    private double targetTemperature = 22.0;
    private boolean heatingOn = false;
    private boolean coolingOn = false;
    private static final double HYSTERESIS = 1.0;

    public TemperatureController(SmartHomeData homeData) {
        this.homeData = homeData;
    }

    @Override
    public void update(String state, Object value) {
        if ("TEMPERATURE".equals(state)) {
            double temp = (Double) value;
            homeData.addReading("TEMPERATURE", temp);

            if (temp < targetTemperature - HYSTERESIS) {
                heatingOn = true;
                coolingOn = false;
            } else if (temp > targetTemperature + HYSTERESIS) {
                heatingOn = false;
                coolingOn = true;
            } else {
                heatingOn = false;
                coolingOn = false;
            }

            System.out.println("TemperatureController: Current temperature " + temp);
            System.out.println("Heater: " + (heatingOn ? "ON" : "OFF") + ", AC: " + (coolingOn ? "ON" : "OFF"));
        }
    }

    public boolean isHeatingOn() {
        return heatingOn;
    }

    public boolean isCoolingOn() {
        return coolingOn;
    }

    public void setTargetTemperature(double temperature) {
        this.targetTemperature = temperature;
    }
}

File: HomeAutomation/HomeAutomation/Components/TemperatureController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using HomeAutomation.Core;

namespace HomeAutomation.Components
{
    public class TemperatureController : IComponent
    {
        public double CurrentTemperature { get; private set; }
        public double TargetTemperature { get; private set; }
        public bool HeatingOn { get; private set; }
        public bool CoolingOn { get; private set; }
        private const double Hysteresis = 0.5;

        public TemperatureController(double targetTemperature)
        {
            TargetTemperature = targetTemperature;
        }

        public void OnStateChanged(EnvironmentalState state)
        {
            CurrentTemperature = state.Temperature;
            if (CurrentTemperature < TargetTemperature - Hysteresis)
            {
                HeatingOn = true;
                CoolingOn = false;
            }
            else if (CurrentTemperature > TargetTemperature + Hysteresis)
            {
                CoolingOn = true;
                HeatingOn = false;
            }
            else
            {
                HeatingOn = false;
                CoolingOn = false;
            }
        }
    }
}
